docs/proposals/caching.adoc

```diff
@@ -0,0 +1,164 @@
+
+=== ods-start and cleanup
+
+Build tasks supporting caching can count on:
+
+* cache cleanups not occurring while they are running
+* cleanups will not be partially completed. 
+
+On the other hand build tasks must not assume that the cache is still available from a prior build. 
+
+The cleanup strategy described here is subject to change even in patch versions and specific details must not be relied on.
+
+Pipeline cleanup happens during `ods-start`
+
+The following locations on the PVC are used:
+
+- `/ws/` for uncached workspaces. These are recognized if none of there build tasks enabled caching.
+- `/.cache-ws/<pipeline-name>/` for cached workspaces.
+- `/.cache/<technology-name>/` for global caches of a particular build technology. The technology-name would be defined by the build script.
+
+Before cache cleanup `ods-start` cleans up `/ws/` on the PVC. +
+Next all files directly underneath `/.cache/` which are not directories are deleted. This prevents tasks to forget to define and use a technology-name.
+
+**workspace cache cleanup**
+
+The workspace cache cleanup is skipped if the pipeline has workspace caching disabled to avoid merging workspace cached pipelines into a mainline as describe earlier.
+
+.TODO this could be done by creating the pipeline with `cache-workspace-require-space-mb` forcefully set to 0 or via another flag.
+
+`ods-start` then determines:
+
+- `workspace-required-space-mb-total` :== sum of all declared `cache-workspace-require-space-mb` of build tasks contained in `ods.yaml`
+
+If this is 0 the next step is to look at global cache cleanup some steps below.
+
+Otherwise `ods-start` will determine the size of the workspace directory with a variation of `du /.cache-ws/<pipeline-name>/` and convert the result to megabytes. 
+
+Until there is not enough free space:
+
+- Deleted the oldest workspace. Age should be available via `stat -f '%m' .git/FETCH_HEAD` (see https://stackoverflow.com/a/9229377).
+- If there is none continue below.
+
+If there is still not enough free space:
+
+-  delete `/.cache-ws/<pipeline-name>` if it exists
+-  continue the build in `/ws`, so that its `cwd` is at `/ws` and that files `$WORKING_DIR/.ods/git-dir-commit-sha` will not be created for each of the build tasks.
+
+**global cache cleanup**
+
+`ods-start` determines:
+
+- `cache-required-space-mb-total` :== sum of all declared `cache-global-require-space-mb` of build tasks contained in `ods.yaml`
+
+The global cache cleanup deletes in a similar way as the workspace cache cleaning just described but the cleanup candidates are folders at: `/.cache/<technology-name>/`. 
+
+If there is not sufficient space in the global cache after cleanup the file `.ods/cache-global-parent-dir` will not exist in the workspace directory.
+
+
+== Pro
+
+* Caching enabled for tools using global cache
+
+* Workspace caching enables speedup for technologies which benefit most from an incremental build as opposed to a global dependency cache. 
```

Comment:
maybe we can speed up the build of docker images as well using the global cache.
I assume here that using buildah (https://docs.openshift.com/container-platform/4.9/cicd/builds/custom-builds-buildah.html) part of the docker image layers will be cached. Thou, I'm not aware how the command `oc start-build` works and if it use any kind of caching.

Reply:
Note that `oc start-build` is only used in the Jenkins approach. We do not use `oc` here at all, and rather invoke `buildah` inside the `ods-package-image` task directly (in fact, nothing in ods-pipeline is OpenShift-specific). As mentioned before, I believe `buildah` might be able to use layers stored in the registry as caches, but I have not explored this yet. I would tentatively move this out of the discussion here and address separately.

Reply:
Yes, how to improve image build time is a separate discussion. It would be awesome if we could benefit from the global cache.

Reply:
I have investigated this a bit and create this issue to evaluate how to optimize/improve image creation using local cache: https://github.com/opendevstack/ods-pipeline/issues/433